engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ring/AbstractRingChunkSource.java

```diff
@@ -0,0 +1,389 @@
+import io.deephaven.chunk.ResettableWritableChunk;
+import io.deephaven.chunk.WritableChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.rowset.RowSetFactory;
+import io.deephaven.engine.table.ChunkSource;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.impl.DefaultChunkSource;
+import io.deephaven.engine.table.impl.DefaultGetContext;
+import io.deephaven.util.datastructures.LongRangeConsumer;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Closeable;
+import java.lang.reflect.Array;
+import java.util.Objects;
+import java.util.function.LongConsumer;
+
+import static io.deephaven.engine.table.impl.AbstractColumnSource.USE_RANGES_AVERAGE_RUN_LENGTH;
+
+/**
+ * The base for a ring chunk source. Provides a single contiguous array for ring-buffer operations.
+ *
+ * @param <T> the item type
+ * @param <ARRAY> the array type
+ * @param <SELF> the self type
+ */
+abstract class AbstractRingChunkSource<T, ARRAY, SELF extends AbstractRingChunkSource<T, ARRAY, SELF>>
+        implements DefaultChunkSource<Values> {
+
+    protected final ARRAY ring;
+    protected final int capacity;
+    long nextRingIx;
+
+    private final ResettableWritableChunk<Values> ringView;
+
+    public AbstractRingChunkSource(@NotNull Class<T> componentType, int capacity) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Capacity must be positive");
+        }
+        this.capacity = capacity;
+        // noinspection unchecked
+        ring = (ARRAY) Array.newInstance(componentType, capacity);
+        ringView = getChunkType().makeResettableWritableChunk();
+    }
+
+    /**
+     * The maximum size {@code this} ring can hold. Constant.
+     *
+     * @return the capacity
+     */
+    public final int capacity() {
+        return capacity;
+    }
+
+    /**
+     * The size, {@code 0 <= size <= capacity}. The size will never shrink.
+     *
+     * <p>
+     * Logically equivalent to {@code lastKey - firstKey + 1}.
+     *
+     * @return the size
+     */
+    public final int size() {
+        return capacity <= nextRingIx ? capacity : (int) nextRingIx;
+    }
+
+    /**
+     * {@code true} if empty, else {@code false}. Once {@code false} is returned, will always return {@code false}.
+     *
+     * <p>
+     * Logically equivalent to {@code size == 0}.
+     *
+     * @return {@code true} if empty
+     */
+    public final boolean isEmpty() {
+        return nextRingIx == 0;
+    }
+
+    /**
+     * {@code true} if {@code key} is in the index.
+     *
+     * <p>
+     * Logically equivalent to the condition {@code firstKey <= key <= lastKey}.
+     *
+     * @param key the key
+     * @return {@code true} if {@code key} is in the index.
+     * @see #firstKey()
+     * @see #lastKey()
+     */
+    public final boolean containsKey(long key) {
```

Comment:
See my other comment in `get`. You don't need this method.

Reply:
I think some safety checks *may* be good. See STRICT_KEYS - can turn off if desired.